### PR TITLE
Challenge 3 triage

### LINKED ISSUE #195: SymNats interact poorly with ExprMappable
The interface for `SymNat` maintains the invariant that its inner integer is positive by clamping its value to be necessarily positive.
This conflicts with the expression rebinding infrastructure (i.e. ExprMappable) because the only way to map an expression-modifying function over a 'SymNat' is to: extract its inner integer, apply the function, then wrap the integer back up in a SymNat.

This has a few consequences:

1) Each time a SymNat is transformed (via mapExpr) it adds a new clamp, resulting in terms like `(max 0 (max 0 (max 0 n)))`
2) Applying the identity function to a SymNat still results in a modified internal value.
3) Term equality on SymNats is unstable in this context (without term caching), since each internal int is now formally distinct.

### PR DESCRIPTION
**Pull request by danmatichuk**
Addressing issues that are blocking the verifier from running Challenge 3

See: #130

- Generalizes the functionality from `Main` that determines the architecture from the ELF and loads the PLT symbols from it
    - Fixes: #324 
- Include upstream What4 change that makes converting from and to SymNats less destructive to the term shape.
    - Partially fixes #195 
    - TODO: Remove workaround and update to latest what4 (see #328)
- Move `AssumptionSet` into its own module and collect its interface functions there.
   - Resolves some circular dependencies that were introduced 
   - Used in more places as the standard representation of a collection of assumptions.
   - Explicit bindings/rewrites are useful for simplifications for readability
- Use structured traversal for value domain when widening so that each value domain location is checked separately.
   - Useful for debugging which specific value constraints were causing timeouts 
- Handle cases where Crucible was throwing errors due to inconsistent assumptions
   - Introduced when solver timeout support was added. Occurs in `withSatAssumption` when the incoming assumption
      is trivially false.
- Basic PLT override support
   - Each architecture defines a mapping from PLT symbols to state transformations
   - Not intended to be long term solution. Is sufficient for simple libc calls until we have better override support for crucible.
   - Added PLT stubs for `malloc` and `clock` for Aarch32
   - No PPC support yet (see: https://github.com/GaloisInc/pate/issues/327). Test marked as expected failure.
-  Clean up build system for tests.
   - Factor out duplication between architectures

### CODE REVIEW

File: src/Pate/Config.hs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 {-# LANGUAGE ScopedTypeVariables #-}
 module Pate.Config (
   PatchData(..),
+  PatchDataParseError,
```
Comment (travitch) [inline src/Pate/Config.hs:6]: Should this have its constructors exported too?

File: src/Data/Macaw/CFGSlice.hs [hunk 2]
```diff
@@ -126,15 +128,25 @@ assignBlockEnd archFns blendVar stmt = MSB.crucGenArchConstraints archFns $ do
   blend' <- blockEndAtom archFns blend
   MSB.addStmt $ CR.WriteGlobal blendVar blend'
 
+blockEndCaseEq :: IsSymInterface sym
+               => Proxy arch
```
Comment (travitch) [inline src/Data/Macaw/CFGSlice.hs:132]: It is generally a best practice to make proxies in positive (argument) positions type variables:

```
blockEndCaseEq :: forall sym arch proxy => IsSymInterface sym -> proxy arch -> ...
```

This way the caller can pass in any valid value of the correct kind (given that this function doesn't care what is passed in).

File: src/Pate/Config.hs [hunk 3]
```diff
@@ -202,6 +204,16 @@
...
 
+-- | Controls how abstract domains are shared between function call nodes in the pairgraph.
+data ContextSensitivity =
+    -- | All calls to any given function share the same abstract domain, i.e.
+    --   the resulting domain is proven to hold at every return site.
+    AllSharedAbstractDomains
+    -- | Each call to a function has a distinct abstract domain that holds
+    --   at its (single) return site (distinguished by tagging the function node with
+    --   the calling node)
+  | AllDistinctAbstractDomains
```
Comment (travitch) [inline src/Pate/Config.hs:215]: "All" is not quite correct, is it? What is the key in the map from contexts to abstract values? What happens in the presence of recursion?

File: src/Data/Parameterized/SetF.hs [hunk 4]
```diff
@@ -26,20 +26,34 @@
...
-import Prelude hiding (filter)
+import Prelude hiding (filter, null)
 import           Data.Parameterized.Classes
+import qualified Data.Foldable as Foldable
+
+import qualified Prettyprinter as PP
+import           Prettyprinter ( (<+>) )
+
+import qualified Prettyprinter as PP
+import           Prettyprinter ( (<+>) )
```
Comment (travitch) [inline src/Data/Parameterized/SetF.hs:56]: These imports are duplicated

File: src/Pate/AssumptionSet.hs [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+
+import           Data.Parameterized.SetF ( SetF )
+import qualified Data.Parameterized.SetF as SetF
+import qualified Pate.SimulatorRegisters as PSR
+import           Pate.Panic
+import qualified What4.ExprHelpers as WEH
+import           What4.ExprHelpers ( ExprSet, VarBindCache, ExprBindings, ppExprSet )
+import qualified Pate.ExprMappable as PEM
+
+data AssumptionSet sym =
```
Comment (travitch) [inline src/Pate/AssumptionSet.hs:57]: Is there some kind of invariant that this type is attempting to capture on top of `ExprSet`?

File: src/Pate/ExprMappable.hs [hunk 6]
```diff
@@ -132,3 +140,21 @@ instance ExprMappable sym (SkipTransformation a) where
...
+-- | Deduce ad-hoc 'ExprMappable' instances for 'WI.SymExpr'.
+--   This uses an unsafe coercion to implicitly use the 'ExprMappable' instance
+--   defined in 'ToExprMappable'.
+--   This is necessary because 'WI.SymExpr' is a type family and therefore we cannot
+--   declare the obvious instance for it.
+symExprMappable ::
+  forall sym.
+  sym ->
+  SymExprMappable sym (WI.SymExpr sym)
+symExprMappable _sym = unsafeCoerce r
```
Comment (travitch) [inline src/Pate/ExprMappable.hs:158]: This is quite scary.

Why is this safe? Is there no other way to achieve what you need here?

File: src/What4/ExprHelpers.hs [hunk 7]
```diff
@@ -115,10 +117,69 @@
...
+            let rng_y = W4.integerBounds y
+            W4.unsafeSetAbstractValue (W4AD.rangeMax rng_x rng_y) <$>
+              W4.intIte sym x_le_y y x
+
+-- | Overriding 'W4.integerToNat' to use a modified 'intMax' that
+--   statically evaluates in more cases. In particular this ensures that
+--   integers with a positive abstract domain aren't wrapped in any additional checks,
+--   and that the inner integer is statically known to be positive.
+integerToNat ::
+  W4.IsExprBuilder sym =>
```
Comment (travitch) [inline src/What4/ExprHelpers.hs:170]: I'm obviously always worried about unsafeCoerce. Is there a time horizon to replace this with safe what4 alternatives now that we have improved infrastructure there? If so, can you reference the ticket here?

File: src/Pate/Arch.hs [hunk 8]
```diff
@@ -127,12 +143,80 @@ data ValidArchData arch =
                 --
                 -- For example, these could be PLT stub symbols for ELF binaries
                 , validArchPatchedExtraSymbols :: Map.Map BS.ByteString (BVS.BV (MC.ArchAddrWidth arch))
+                , validArchStubOverrides :: ArchStubOverrides arch
                 }
 
+-- | A PLT stub is allowed to make arbitrary modifications to the symbolic state
```
Comment (travitch) [inline src/Pate/Arch.hs:149]: This is currently used for PLT stubs, but could in principle be used for any function that is to be overridden. That doesn't really require any changes, but might be worth a note in the comment.

File: src/Pate/Arch.hs [hunk 9]
```diff
@@ -127,12 +143,80 @@
...
+mkMallocOverride _rLen rOut = StubOverride $ \sym st -> do
+  let mr = PS.simMaxRegion st
+  let w = MC.memWidthNatRepr @(MC.ArchAddrWidth arch)
+  mr_nat <- W4.integerToNat sym (PS.unSE mr)
+  zero <- W4.bvLit sym w (BVS.mkBV w 2)
+  let fresh_ptr = PSR.ptrToEntry (CLM.LLVMPointer mr_nat zero)
+  mr_inc <- PS.forScopedExpr sym mr $ \sym' mr' -> do
+    one <- W4.intLit sym' 1
+    W4.intAdd sym' mr' one
+  return (st { PS.simMaxRegion = mr_inc, PS.simRegs = ((PS.simRegs st) & (MC.boundValue rOut) .~ fresh_ptr) })
```
Comment (travitch) [inline src/Pate/Arch.hs:183]: Can you add a comment that describes the semantics of the override, perhaps with a justification for the behavior?

File: src/Pate/Discovery.hs [hunk 10]
```diff
@@ -269,8 +256,9 @@ associateFrames ::
   forall sym arch v.
   SimBundle sym arch v ->
   MCS.MacawBlockEndCase ->
-  EquivM sym arch (PSS.AssumptionSet sym v)
-associateFrames bundle exitCase = PPa.catBins $ \get -> do
+  Bool ->
```
Comment (travitch) [inline src/Pate/Discovery.hs:259]: At the very least we should document what this boolean is. Better, we can make it into an ADT that is descriptive at the call site

File: src/Pate/Arch.hs [hunk 11]
```diff
@@ -127,12 +143,80 @@
...
+  16 <= MC.ArchAddrWidth arch =>
+  MS.SymArchConstraints arch =>
+  MC.ArchReg arch (MT.BVType (MC.ArchAddrWidth arch)) {- ^ return register -} ->
+  StubOverride arch
+mkClockOverride rOut = StubOverride $ \sym st -> do
+  let w = MC.memWidthNatRepr @(MC.ArchAddrWidth arch)
+  zero_bv <- W4.bvLit sym w (BVS.mkBV w 0)
+  zero_nat <- W4.natLit sym 0
+  let zero_ptr = PSR.ptrToEntry (CLM.LLVMPointer zero_nat zero_bv)
+  return (st { PS.simRegs = ((PS.simRegs st) & (MC.boundValue rOut) .~ zero_ptr) })
```
Comment (travitch) [inline src/Pate/Arch.hs:196]: Likewise, can you document the semantics of the override?

File: src/Pate/Equivalence.hs [hunk 12]
```diff
@@ -444,27 +449,29 @@ regCondToAsm ::
...
   RegisterCondition sym arch v ->
-  IO (AssumptionSet sym v)
+  IO (AssumptionSet sym)
 regCondToAsm _sym regCond = return $ PRt.collapse (regCondPreds regCond)
 
 -- | A structured pre condition
 data StatePreCondition sym arch v = StatePreCondition
   { stRegPreCond :: RegisterCondition sym arch v
   , stStackPreDom :: PEM.MemoryDomain sym arch
   , stMemPreDom :: PEM.MemoryDomain sym arch
```
Comment (travitch) [inline src/Pate/Equivalence.hs:459]: What should I (the reader) expect to appear in an "extra" assumption set?

File: src/Pate/Equivalence.hs [hunk 13]
```diff
@@ -476,13 +483,20 @@
...
-eqDomPre sym inO inP (EquivContext hdr _stackRegion) eqDom  = do
+eqDomPre sym inO inP (eqCtxHDR -> hdr) eqDom  = do
   let
     stO = simInState inO
     stP = simInState inP
     
   regsEq <- regDomRel hdr sym stO stP (PED.eqDomainRegisters eqDom)
-  return $ StatePreCondition regsEq (PED.eqDomainStackMemory eqDom) (PED.eqDomainGlobalMemory eqDom)
+  -- TODO: we haven't included this in the domain, and therefore we're simply
+  -- requiring that the number of allocations for each program always match
```
Comment (travitch) [inline src/Pate/Equivalence.hs:493]: Can you open an issue to track this?

File: src/Pate/AssumptionSet.hs [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+macawRegBinding _sym var val = do
+  case PSR.macawRegRepr var of
+    CLM.LLVMPointerRepr _ -> ptrBinding (PSR.macawRegValue var) (PSR.macawRegValue val)
+    CT.BoolRepr -> exprBinding (PSR.macawRegValue var) (PSR.macawRegValue val)
+    CT.StructRepr Ctx.Empty -> mempty
+    _ -> panic Rewriter "macawRegBinding" ["Unexpected macaw types"
+                                          , show (PSR.macawRegRepr var)
+                                          , show (PSR.macawRegRepr val)
+                                          ]
+
```
Comment (travitch) [inline src/Pate/AssumptionSet.hs:197]: These would all benefit from saying something about what the meaning of the result is (and hopefully that is in terms of the description of `AssumptionSet`)

File: src/Pate/AssumptionSet.hs [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+  m f
+apply sym asm f = do
+  cache <- IO.liftIO WEH.freshVarBindCache
+  apply' sym cache asm f
+
+-- | Rewrite the given 'f' with any bindings in the given 'AssumptionSet'.
+--   Bindings are applied repeatedly to each component expression until a fixpoint
+--   is reached or a loop is detected.
+apply' ::
+  forall sym f m t solver fs.
```
Comment (travitch) [inline src/Pate/AssumptionSet.hs:220]: Do both of these need to be exported?

File: src/Pate/AssumptionSet.hs [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+  PEM.ExprMappable sym f =>
+  IO.MonadIO m =>
+  sym ~ (W4B.ExprBuilder t solver fs) =>
+  sym ->
+  VarBindCache sym ->
+  AssumptionSet sym ->
+  f ->
+  m f
+apply' sym cache asm f = do
+  let
```
Comment (travitch) [inline src/Pate/AssumptionSet.hs:230]: If this does need to be exported, a more descriptive name like `applyWithCache` would be better than `apply'`---understanding what the prime means in isolation is difficult

File: src/Pate/AssumptionSet.hs [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+  cache <- IO.liftIO WEH.freshVarBindCache
+  origAsm' <- PEM.mapExpr sym (apply' sym cache newAsm) origAsm
+  return $ newAsm <> origAsm'
+
+-- | Retrieve a value that the given expression is bound to in
+--   the given 'AssumptionSet'.
+--   In the case of multiple results, constant values are preferred,
+--   otherwise the least element of all possible symbolic bindings is returned.
+getSomeBinding ::
+  forall sym tp.
```
Comment (travitch) [inline src/Pate/AssumptionSet.hs:260]: Can you say something about why one might want this?

File: src/Pate/AssumptionSet.hs [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+      in SetF.unions (map (\e' -> SetF.insert e' (getAllBindings sym ancestors' asm e')) (SetF.toList es))
+    Nothing -> case W4.exprType e of
+      W4.BaseBoolRepr | isAssumedPred asm e -> SetF.singleton (W4.truePred sym)
+      _ -> SetF.empty
+
+-- | Compute a predicate that collects the individual assumptions in the frame, including
+-- equality on all bindings.
+toPred ::
+  forall sym m.
+  W4.IsSymExprBuilder sym =>
```
Comment (travitch) [inline src/Pate/AssumptionSet.hs:296]: Can you state precisely when the returned predicate is true? Is it true if all of the assumptions hold if you assume the initial equalities on bindings?

**Comment by travitch** [comment]
I did my best here given the size. If there is anything you'd like me to focus on more closely, let me know and I will make an attempt

File: src/Pate/AssumptionSet.hs (hunk 15 above)
Comment (danmatichuk) [reply src/Pate/AssumptionSet.hs:220]: At the moment, no, but I had some ambitions to use it later.

File: src/Pate/AssumptionSet.hs (hunk 5 above)
Comment (danmatichuk) [reply src/Pate/AssumptionSet.hs:72]: I don't think there are any explicit invariants here. The assumption set is a collection of predicates for general predicates as well as bindings used for equality assumptions that can be applied to perform simplifications.

The state of this binding environment may need to be revisited at some point to ensure that expressions can be consistently simplified into some kind of normal form (which would imply some kind of invariant on what shape the bindings in the environment are allowed to have). At the moment the rewriter that consumes the bindings doesn't make any assumptions about their shape (i.e. it implicitly breaks rewrite loops while trying to find a rewriting fixpoint).

**Pull request merged**